ebpf/throughput/throughput.c

```diff
@@ -0,0 +1,95 @@
+#include "../headers/bpf.h"
+
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+
+#include <linux/if_ether.h>
+#include <linux/ip.h>
+#include <linux/ipv6.h>
+#include <linux/pkt_cls.h>
+#include <linux/tcp.h>
+#include <linux/udp.h>
+
+#include <arpa/inet.h>
+
+char __license[] SEC("license") = "Dual MIT/GPL";
+
+struct throughput_event
+{
+	__u64 timestamp;
+	__u32 size;
+	__u16 sport;
+	__u16 dport;
+	__u8  saddr[16];
+	__u8  daddr[16];
+};
+
+// used for sending throughput events to userspace
+struct {
+    __uint(type, BPF_MAP_TYPE_RINGBUF);
+    __uint(max_entries, 1 << 16);
+} throughput_events SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__type(key, __u32);
+	__type(value, struct throughput_event);
+	__uint(max_entries, 1);
+} throughput_event_heap SEC(".maps");
+
+SEC("classifier")
+int packet_classifier(struct __sk_buff *skb) {
+	void *data = (void*)(long)skb->data;
+	void *data_end = (void*)(long)skb->data_end;
+
+	if (data + sizeof(struct ethhdr) > data_end) {
+    	return TC_ACT_UNSPEC;
+    }
+
+	// Get the Ethernet header
+	struct ethhdr *eth = data;
+
+	if (eth->h_proto == htons(ETH_P_IP)) {
+		if (data + sizeof(struct ethhdr) + sizeof(struct iphdr) > data_end) {
+			return TC_ACT_UNSPEC;
+		}
+		struct iphdr *ip = data + sizeof(struct ethhdr);
+
+		struct throughput_event *e = bpf_ringbuf_reserve(&throughput_events, sizeof(struct throughput_event), 0);
```

Comment:
Here we "reserve" a throughput event on the ringbuffer, which we will write back to it later on. The verifier requires us to submit the struct again before we return, so even if we are unable to read port numbers from the TCP or UDP header, we write it back anyway.